<launch>
  <arg name="voxel_size" default="0.08"/>
  <arg name="pc_pub_rate" default="10"/>
  <arg name="adaptive_mapping" default="false"/>
  <arg name="crf_only_on_updated_voxels" default="true"/>
  <arg name="collection_name" default="collection1" />
  <arg name="use_gt_semantic" default="true" />
  <arg name="truncation_distance" default="0.2" />
  

  <!-- Download the dataset here: https://projects.asl.ethz.ch/datasets/doku.php?id=iros2017 -->

  <node name="voxblox_node" pkg="map_adapt_ros" type="adaptive_voxel_server" output="screen" args="-alsologtostderr" clear_params="true" required="true">
    <remap from="pointcloud" to="/traj_visualizer/pcl_output"/>
    <remap from="transform" to="/traj_visualizer/tf_output" />
    <remap from="/commander" to="/traj_visualizer/commander" />
    <param name="tsdf_voxel_size" value="$(arg voxel_size)" />
    <param name="truncation_distance" value="$(arg truncation_distance)" />
    <!-- <param name="start_voxel_subsampling_factor" value="4.0" /> -->
    <!-- <param name="max_weight" value="5.0" /> -->
    <param name="tsdf_voxels_per_side" value="16" />
    <param name="voxel_carving_enabled" value="false" />
    <param name="color_mode" value="color" />
    <param name="use_tf_transforms" value="false" />
    <param name="update_mesh_every_n_sec" value="-2" />
    <param name="min_time_between_msgs_sec" value="0.0" />
    <param name="method" value="fast" />
    <param name="use_const_weight" value="simple" />
    <param name="allow_clear" value="false" />
    <param name="verbose" value="true" />
    <param name="min_ray_length_m" value="0.0" />
    <param name="max_ray_length_m" value="10.0" />
    <param name="visualize_semantic" value="false" />
    <param name="max_consecutive_ray_collisions" value="3" />
    <param name="load_semantic_probability" value="true" />
    <param name="load_geo_complexity" value="true" />
    <param name="adaptive_mapping" value="$(arg adaptive_mapping)" />
    <param name="adaptive_ratio_small" value="8.0" />
    <param name="adaptive_ratio_middle" value="2.0" />
    <param name="fuse_semantic_for_subvoxel" value="true" />
    <param name="intensity_colormap" value="nyu" />
    <param name="intensity_max_value" value="40.0" />
    <param name="semantic_update_method" value="bayesian" /> <!--Support: max_pooling; weighted_max; bayesian; bayesian_constant-->
    <rosparam param="small_semantic" >[3, 6, 12, 14, 16, 18, 19, 21, 23, 31, 34, 35, 37, 39]</rosparam> 
    <rosparam param="large_semantic" >[1, 2, 4, 7, 8, 11, 22, 25, 27, 28, 32, 36]</rosparam>
    <!-- <rosparam param="small_semantic" >[2, 3, 6, 8, 10, 13, 16, 22, 23, 30, 33, 34, 37, 38]</rosparam> 
    <rosparam param="large_semantic" >[7, 9, 17, 24, 25, 27, 35, 39]</rosparam> -->
    <rosparam param="geo_thresholds" >[0.05,0.1]</rosparam> 
    <rosparam file="$(find map_adapt_ros)/cfg/identity.yaml"/>
    <param name="mesh_filename" value="$(find map_adapt_ros)/mesh_results/jianhao_habitat.ply" />
    <param name="scene_name" value="$(arg collection_name)" />
    <param name="semantic_level_split_id" value="4" />
  </node>

  <node name="traj_visualizer" pkg="build_map" type="pub_pc_cvpr_scannet_geo" output="screen" args="-alsologtostderr" clear_params="true">
    <remap from="/traj_visualizer/pub_pc_switch" to="/voxblox_node/command_to_pc_publisher"/>
    <param name="root_dir" value="/media/jianhao/T7/MAP_ADAPT/data/Scannet/$(arg collection_name)/" />
    <param name="path_to_estimated_pose" value="est_pose.txt" />
    <param name="rate" value="$(arg pc_pub_rate)" />
    <param name="wait_time" value="3.0" />
    <param name="use_gt" value="$(arg use_gt_semantic)" />
  </node>

</launch>
